File: apps/sensor-display-as/assembly/sht31.ts
// SHT31 / SHT30 の読み出し。apps/README.md §1 が正。

import { I2cBus, time } from "../../../bindings/assemblyscript/assembly/index";

/// I2C アドレス。
export const ADDRESS: u16 = 0x44;

/// 計測が終わるまでの待ち時間。
const MEASURE_MS: u32 = 15;

/// CRC-8。多項式 0x31、初期値 0xFF、反転なし。
export function crc8(data: Uint8Array, from: i32, len: i32): u8 {
  let crc: u8 = 0xff;
  for (let i = 0; i < len; i++) {
    crc ^= data[from + i];
    for (let b = 0; b < 8; b++) {
      crc = (crc & 0x80) != 0 ? <u8>((crc << 1) ^ 0x31) : <u8>(crc << 1);
    }
  }
  return crc;
}

/// 読み出しの結果。`ok` が false のときは `crcFailed` で理由を分ける。
export class Reading {
  ok: bool = false;
  crcFailed: bool = false;
  rawT: u16 = 0;
  rawH: u16 = 0;
}

const MEASURE = new Uint8Array(2);
const FRAME = new Uint8Array(6);

/// 1 回測って読む。
export function read(bus: I2cBus): Reading {
  const r = new Reading();
  MEASURE[0] = 0x24;
  MEASURE[1] = 0x00;
  if (bus.write(ADDRESS, MEASURE) != 0) return r;
  time.sleepMs(MEASURE_MS);

  if (bus.read(ADDRESS, FRAME) != 6) return r;
  if (crc8(FRAME, 0, 2) != FRAME[2] || crc8(FRAME, 3, 2) != FRAME[5]) {
    r.crcFailed = true;
    return r;
  }
  r.ok = true;
  r.rawT = <u16>((<u16>FRAME[0] << 8) | <u16>FRAME[1]);
  r.rawH = <u16>((<u16>FRAME[3] << 8) | <u16>FRAME[4]);
  return r;
}

/// 温度（摂氏 ×100）。固定小数で計算する。
export function tempCenti(rawT: u16): i32 {
  return -4500 + (17500 * <i32>rawT) / 65535;
}

/// 相対湿度（％ ×100）。
export function humidityCenti(rawH: u16): i32 {
  return (10000 * <i32>rawH) / 65535;
}

File: apps/sensor-display-rs/src/sht31.rs
//! SHT31 / SHT30 の読み出し。`apps/README.md` §1 が正。

use wasmicon_hal::i2c::Bus;
use wasmicon_hal::time;

/// I2C アドレス。
pub const ADDRESS: u16 = 0x44;

/// 単発計測（高精度・クロックストレッチなし）。
const MEASURE: [u8; 2] = [0x24, 0x00];

/// 計測が終わるまでの待ち時間。
const MEASURE_MS: u32 = 15;

/// CRC-8。多項式 0x31、初期値 0xFF、反転なし。
#[must_use]
pub fn crc8(data: &[u8]) -> u8 {
    let mut crc: u8 = 0xFF;
    for &b in data {
        crc ^= b;
        let mut i = 0;
        while i < 8 {
            crc = if crc & 0x80 != 0 {
                (crc << 1) ^ 0x31
            } else {
                crc << 1
            };
            i += 1;
        }
    }
    crc
}

/// 生の測定値。
pub struct Reading {
    pub raw_t: u16,
    pub raw_h: u16,
}

/// 読み出しの失敗。
pub enum Error {
    /// 通信できない。
    Io,
    /// CRC が合わない。
    Crc,
}

/// 1 回測って読む。
///
/// # Errors
/// 通信に失敗したら `Io`、CRC が合わなければ `Crc`。
pub fn read(bus: &Bus) -> Result<Reading, Error> {
    if bus.write(ADDRESS, &MEASURE).is_err() {
        return Err(Error::Io);
    }
    time::sleep_ms(MEASURE_MS);

    let mut buf = [0u8; 6];
    match bus.read(ADDRESS, &mut buf) {
        Ok(6) => {}
        _ => return Err(Error::Io),
    }
    if crc8(&buf[0..2]) != buf[2] || crc8(&buf[3..5]) != buf[5] {
        return Err(Error::Crc);
    }
    Ok(Reading {
        raw_t: (u16::from(buf[0]) << 8) | u16::from(buf[1]),
        raw_h: (u16::from(buf[3]) << 8) | u16::from(buf[4]),
    })
}

/// 温度（摂氏 ×100）。固定小数で計算する。
#[must_use]
pub fn temp_centi(raw_t: u16) -> i32 {
    -4500 + (17500 * i32::from(raw_t)) / 65535
}

/// 相対湿度（％ ×100）。
#[must_use]
pub fn humidity_centi(raw_h: u16) -> i32 {
    (10000 * i32::from(raw_h)) / 65535
}
